### DISCUSSION TITLE
Got ModuleNotFoundError: No module named 'cmake' when building on windows

### DISCUSSION THREAD

**Discussion by 1yefuwang1** [pypa/cibuildwheel#1901] [Q&A] [answered]
Hi comunity, I'm trying to figure out a strange error which only appears on WIndows, not on linux and macos(succeeded github actions run [here](https://github.com/1yefuwang1/vectorlite/actions/runs/9725362412)).
The failed github action is [here](https://github.com/1yefuwang1/vectorlite/actions/runs/9725542058/job/26842986199).
The workflow file is [here](https://github.com/1yefuwang1/vectorlite/blob/win_ci/.github/workflows/build-wheel.yml)

The most relevant error message is:
```
       Traceback (most recent call last):
          File "C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\lib\runpy.py", line 196, in _run_module_as_main
            return _run_code(code, main_globals, None,
          File "C:\Users\runneradmin\AppData\Local\pypa\cibuildwheel\Cache\nuget-cpython\python.3.10.11\tools\lib\runpy.py", line 86, in _run_code
            exec(code, run_globals)
          File "C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-2jflgb46\overlay\Scripts\cmake.exe\__main__.py", line 4, in <module>
        ModuleNotFoundError: No module named 'cmake'
```
In my [setup.py](https://github.com/1yefuwang1/vectorlite/blob/win_ci/setup.py#L48) script, cmake is installed using pip at `C:\\Users\\runneradmin\\AppData\\Local\\Temp\\pip-build-env-2jflgb46\\overlay\\Lib\\site-packages\\cmake\\data\\bin\\cmake` and run using `subprocess.run`, but the error message points to `"C:\Users\runneradmin\AppData\Local\Temp\pip-build-env-2jflgb46\overlay\Scripts\cmake.exe\__main__.py"`, which is confusing.

I tried to manually install cmake to PATH (based on a similar question https://stackoverflow.com/questions/74226651/error-in-building-wheel-for-osqp-because-no-module-named-cmake), but it still doesn't work.

**Comment by henryiii** [comment, marked as answer]
A few quick comments, will try to look at it later:

I'd highly recommend looking at scikit-build-core, it's designed to build CMake extension modules. Your approach has several issues. You shouldn't add `cmake` and `ninja` to your build system requirements, as systems like Conda, Linux package managers, etc. do not supply these modules. And you should never import `cmake` without having a fallback on a system-installed CMake. Scikit-build-core does this for you correctly. Also, you are using `wheel.bdist_wheel`, which is deprecated (wheel has never claimed to have a public API), and is being moved to setuptools in the next release, and then removed from wheel entirely. Again, scikit-build-core fixes this by not using setuptools at all. :)

That said, the problem here is that it seems to be muddling a `cmake` entrypoint with an environment that it wasn't installed with. I'll have to try to look and see why that happens later.

**Reply by henryiii** [reply]
vcpackage is trying to run CMake, and it’s not using the correct environment. The CMake entry point can’t be used if CMake is not importable. The outer CMake run is fine.

**Reply by 1yefuwang1** [reply]
Thank you for the reply. I'll check out scikit-build-core. It looks a good fit for my project except it doesn't seem to support `cmake preset` yet.

About adding `cmake` and `ninja` to build system requirements, I expect them to be installed using pip(instead of Conda, linux package managers) when building wheels, which seems to work fine , except that vcpkg doesn't invoke cmake with correct settings.

**Reply by 1yefuwang1** [reply]
This issue is [resolved](https://github.com/1yefuwang1/vectorlite/pull/2/files#diff-a8f0a3007b7fec81522975242e4687e7812cbea1bc5d5fcba166665cfa74db3b) by using system installed `cmake` instead of relying on pip-installed one.
Good to know the deprecation of `wheel.bdist_wheel`. Will investigate more about Scikit-build-core.

**Reply by henryiii** [reply]
Opened an issue here: https://github.com/scikit-build/scikit-build-core/issues/790

**Reply by henryiii** [reply]
Thought it should be possible already with `cmake.args = ["--preset=default"]`.
